File: tests/fixtures/sample.go
// Package store implements an in-memory session store with optional
// write-through to a backing cache. It is safe for concurrent use; all
// exported methods take the package-level mutex.
package store

import (
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/example/cache"
)

// ErrNotFound is returned when a session id is absent from the store.
var ErrNotFound = errors.New("store: session not found")

// ErrExpired is returned when a session exists but its TTL has elapsed.
var ErrExpired = errors.New("store: session expired")

// DefaultTTL is the fallback session lifetime when none is supplied.
const DefaultTTL = 30 * time.Minute

// maxSessions caps the number of live sessions before eviction kicks in.
const maxSessions = 4096

// Session is a single user session held by the store.
type Session struct {
	ID        string
	UserID    int64
	CreatedAt time.Time
	ExpiresAt time.Time
	Data      map[string]string
}

// Repository is the behaviour the store exposes to its callers.
type Repository interface {
	Get(id string) (*Session, error)
	Save(s *Session) error
	Delete(id string) error
	Len() int
}

// Store is the concrete in-memory Repository implementation.
type Store struct {
	mu       sync.RWMutex
	sessions map[string]*Session
	cache    *cache.Client
	ttl      time.Duration
	now      func() time.Time
}

// Option configures a Store at construction time.
type Option func(*Store)

// WithTTL overrides the default session TTL.
func WithTTL(d time.Duration) Option {
	return func(s *Store) {
		s.ttl = d
	}
}

// WithCache attaches a write-through cache client.
func WithCache(c *cache.Client) Option {
	return func(s *Store) {
		s.cache = c
	}
}

// New constructs a Store with the supplied options applied.
func New(opts ...Option) *Store {
	s := &Store{
		sessions: make(map[string]*Session),
		ttl:      DefaultTTL,
		now:      time.Now,
	}
	for _, opt := range opts {
		opt(s)
	}
	return s
}

// Get returns the session for id or ErrNotFound / ErrExpired.
func (s *Store) Get(id string) (*Session, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	sess, ok := s.sessions[id]
	if !ok {
		return nil, ErrNotFound
	}
	if s.isExpired(sess) {
		return nil, ErrExpired
	}
	return sess, nil
}

// Save inserts or replaces a session, evicting if at capacity.
func (s *Store) Save(sess *Session) error {
	if sess == nil || sess.ID == "" {
		return fmt.Errorf("store: invalid session")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.sessions) >= maxSessions {
		s.evictOldest()
	}
	if sess.ExpiresAt.IsZero() {
		sess.ExpiresAt = s.now().Add(s.ttl)
	}
	s.sessions[sess.ID] = sess
	if s.cache != nil {
		go s.writeThrough(sess)
	}
	return nil
}

// Delete removes a session if present; absence is not an error.
func (s *Store) Delete(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.sessions, id)
	return nil
}

// Len reports the number of sessions currently held.
func (s *Store) Len() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.sessions)
}

// Purge removes all expired sessions and returns the count removed.
func (s *Store) Purge() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	removed := 0
	for id, sess := range s.sessions {
		if s.isExpired(sess) {
			delete(s.sessions, id)
			removed++
		}
	}
	return removed
}

// isExpired reports whether sess has passed its expiry instant.
func (s *Store) isExpired(sess *Session) bool {
	return !sess.ExpiresAt.IsZero() && s.now().After(sess.ExpiresAt)
}

// evictOldest drops the session with the earliest CreatedAt.
func (s *Store) evictOldest() {
	var oldestID string
	var oldest time.Time
	for id, sess := range s.sessions {
		if oldestID == "" || sess.CreatedAt.Before(oldest) {
			oldestID = id
			oldest = sess.CreatedAt
		}
	}
	if oldestID != "" {
		delete(s.sessions, oldestID)
	}
}

// writeThrough mirrors a session into the attached cache, best-effort.
func (s *Store) writeThrough(sess *Session) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("store: writeThrough recovered:", r)
		}
	}()
	key := fmt.Sprintf("sess:%s", sess.ID)
	_ = s.cache.Set(key, sess.UserID, s.ttl)
}

// NewSession builds a Session with a fresh creation timestamp.
func NewSession(id string, userID int64) *Session {
	return &Session{
		ID:        id,
		UserID:    userID,
		CreatedAt: time.Now(),
		Data:      make(map[string]string),
	}
}

// MergeData copies src entries into dst, returning the mutated dst.
func MergeData(dst, src map[string]string) map[string]string {
	if dst == nil {
		dst = make(map[string]string)
	}
	for k, v := range src {
		dst[k] = v
	}
	return dst
}

// validateID enforces the internal id format (private helper).
func validateID(id string) error {
	if len(id) < 8 {
		return errors.New("store: id too short")
	}
	for _, r := range id {
		if r == ' ' {
			return errors.New("store: id contains space")
		}
	}
	return nil
}

// countActive is a private helper that tallies non-expired sessions.
func countActive(s *Store) int {
	n := 0
	for _, sess := range s.sessions {
		if !s.isExpired(sess) {
			n++
		}
	}
	return n
}

// dumpStats is a private diagnostic that prints store internals.
func dumpStats(s *Store) string {
	return fmt.Sprintf("sessions=%d ttl=%s", len(s.sessions), s.ttl)
}

// Stats reports counts of total and active sessions for callers.
func (s *Store) Stats() (total, active int) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.sessions), countActive(s)
}

File: tests/fixtures/sample.py
"""User session store with optional write-through caching.

This module is safe for concurrent use within a single process; the
public API guards mutable state with a re-entrant lock.
"""
from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from typing import Callable, Optional


DEFAULT_TTL = 30 * 60
MAX_SESSIONS = 4096


class StoreError(Exception):
    """Base class for all store errors."""


class NotFoundError(StoreError):
    """Raised when a session id is absent from the store."""


class ExpiredError(StoreError):
    """Raised when a session exists but its TTL has elapsed."""


@dataclass
class Session:
    """A single user session held by the store."""

    id: str
    user_id: int
    created_at: float = field(default_factory=time.time)
    expires_at: float = 0.0
    data: dict = field(default_factory=dict)

    def is_expired(self, now: float) -> bool:
        """Report whether this session has passed its expiry instant."""
        return self.expires_at != 0.0 and now > self.expires_at

    def touch(self, ttl: float, now: float) -> None:
        """Extend this session's expiry by ttl seconds from now."""
        self.expires_at = now + ttl


class Store:
    """In-memory session store with optional write-through cache."""

    def __init__(
        self,
        ttl: float = DEFAULT_TTL,
        cache: Optional[object] = None,
        now: Callable[[], float] = time.time,
    ) -> None:
        self._sessions: dict[str, Session] = {}
        self._lock = threading.RLock()
        self._ttl = ttl
        self._cache = cache
        self._now = now

    def get(self, sid: str) -> Session:
        """Return the session for sid or raise NotFound / Expired."""
        with self._lock:
            sess = self._sessions.get(sid)
            if sess is None:
                raise NotFoundError(sid)
            if sess.is_expired(self._now()):
                raise ExpiredError(sid)
            return sess

    def save(self, sess: Session) -> None:
        """Insert or replace a session, evicting if at capacity."""
        if not sess.id:
            raise StoreError("invalid session")
        with self._lock:
            if len(self._sessions) >= MAX_SESSIONS:
                self._evict_oldest()
            if sess.expires_at == 0.0:
                sess.touch(self._ttl, self._now())
            self._sessions[sess.id] = sess
            if self._cache is not None:
                self._write_through(sess)

    def delete(self, sid: str) -> None:
        """Remove a session if present; absence is not an error."""
        with self._lock:
            self._sessions.pop(sid, None)

    def __len__(self) -> int:
        with self._lock:
            return len(self._sessions)

    def purge(self) -> int:
        """Remove all expired sessions; return the count removed."""
        removed = 0
        with self._lock:
            now = self._now()
            for sid in list(self._sessions):
                if self._sessions[sid].is_expired(now):
                    del self._sessions[sid]
                    removed += 1
        return removed

    def stats(self) -> tuple[int, int]:
        """Return (total, active) session counts."""
        with self._lock:
            now = self._now()
            active = sum(
                1 for s in self._sessions.values() if not s.is_expired(now)
            )
            return len(self._sessions), active

    def _evict_oldest(self) -> None:
        """Drop the session with the earliest created_at (private)."""
        oldest_id = None
        oldest = float("inf")
        for sid, sess in self._sessions.items():
            if sess.created_at < oldest:
                oldest = sess.created_at
                oldest_id = sid
        if oldest_id is not None:
            del self._sessions[oldest_id]

    def _write_through(self, sess: Session) -> None:
        """Mirror a session into the attached cache, best-effort."""
        try:
            self._cache.set(f"sess:{sess.id}", sess.user_id, self._ttl)
        except Exception:
            pass


def new_session(sid: str, user_id: int) -> Session:
    """Build a Session with a fresh creation timestamp."""
    return Session(id=sid, user_id=user_id)


def merge_data(dst: Optional[dict], src: dict) -> dict:
    """Copy src entries into dst, returning the mutated dst."""
    if dst is None:
        dst = {}
    for key, value in src.items():
        dst[key] = value
    return dst


def build_store(ttl: float = DEFAULT_TTL, cache: Optional[object] = None) -> Store:
    """Construct a Store, wiring up the supplied cache if present."""
    store = Store(ttl=ttl, cache=cache)
    return store


def _validate_id(sid: str) -> None:
    """Enforce the internal id format (private helper)."""
    if len(sid) < 8:
        raise StoreError("id too short")
    if " " in sid:
        raise StoreError("id contains space")


def _count_active(store: Store) -> int:
    """Tally non-expired sessions (private helper)."""
    total, active = store.stats()
    return active


def _dump_stats(store: Store) -> str:
    """Render store internals for diagnostics (private helper)."""
    total, active = store.stats()
    return f"sessions={total} active={active}"


def with_retry(times: int = 3):
    """Decorator: retry the wrapped callable up to `times` times."""

    def decorator(fn: Callable) -> Callable:
        def wrapper(*args, **kwargs):
            last = None
            for _ in range(times):
                try:
                    return fn(*args, **kwargs)
                except Exception as exc:  # noqa: BLE001
                    last = exc
            if last is not None:
                raise last

        return wrapper

    return decorator


@with_retry(times=2)
def flush_store(store: Store) -> int:
    """Purge a store, retrying transient failures."""
    return store.purge()


class Metrics:
    """Lightweight counter bag for store observability."""

    def __init__(self) -> None:
        self.hits = 0
        self.misses = 0

    def record_hit(self) -> None:
        """Increment the hit counter."""
        self.hits += 1

    def record_miss(self) -> None:
        """Increment the miss counter."""
        self.misses += 1

    def ratio(self) -> float:
        """Return the hit ratio, or 0.0 when no observations exist."""
        total = self.hits + self.misses
        if total == 0:
            return 0.0
        return self.hits / total
